PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+            let hasOriginalPurchaseDate = maybeCachedPurchaserInfo != nil
+                                          && maybeCachedPurchaserInfo!.originalPurchaseDate != nil
+
+            let receiptHasTransactions = self.receiptParser.receiptHasTransactions(receiptData: receiptData)
+
+            if !receiptHasTransactions && hasOriginalPurchaseDate {
+                if let completion = maybeCompletion {
+                    self.operationDispatcher.dispatchOnMainThread {
+                        completion(maybeCachedPurchaserInfo, nil)
+                    }
+                }
+                return
+            }
+
+            self.backend.post(receiptData: receiptData,
+                              appUserID: currentAppUserID,
+                              isRestore: isRestore,
+                              productInfo: nil,
+                              presentedOfferingIdentifier: nil,
+                              observerMode: !self.finishTransactions,
+                              subscriberAttributes: unsyncedAttributes) { maybePurchaserInfo, maybeError in
+                self.handleReceiptPost(withPurchaserInfo: maybePurchaserInfo,
+                                       error: maybeError,
+                                       subscriberAttributes: unsyncedAttributes,
+                                       completion: maybeCompletion)
+            }
+        }
+    }
+
+    func handleReceiptPost(withPurchaserInfo maybePurchaserInfo: PurchaserInfo?,
+                           error maybeError: Error?,
+                           subscriberAttributes: SubscriberAttributeDict,
+                           completion maybeCompletion: ReceivePurchaserInfoBlock?) {
+        operationDispatcher.dispatchOnMainThread {
+            if let purchaserInfo = maybePurchaserInfo {
+                self.purchaserInfoManager.cache(purchaserInfo: purchaserInfo, appUserID: self.appUserID)
+            }
+
+            self.markSyncedIfNeeded(subscriberAttributes: subscriberAttributes,
+                                    appUserID: self.appUserID,
+                                    maybeError: maybeError)
+
+            if let completion = maybeCompletion {
+                self.operationDispatcher.dispatchOnMainThread {
+                    completion(maybePurchaserInfo, maybeError)
+                }
+            }
+        }
+    }
+
+    // Although both SKProduct.productIdentifier and SKPayment.productIdentifier
+    // are supposed to be non-null, we've seen instances where this is not true.
+    // so we cast into optionals in order to check nullability, and try to fall back if possible.
+    func extractProductIdentifier(fromProduct product: SKProduct, orPayment payment: SKPayment) -> String? {
+        if let identifierFromProduct = product.productIdentifier as String?,
+           !identifierFromProduct.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
```

Comment:
In Swift, we get an empty string from `SKProduct().productIdentifier`, in ObjC it is `nil`. Like wtf.